echo-webhooks/src/main/groovy/com/netflix/spinnaker/echo/scm/BitbucketWehbookEventHandler.java

```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright 2019 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.echo.scm;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.netflix.spinnaker.echo.model.Event;
+import lombok.Data;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.Map;
+
+@Component
+public class BitbucketWehbookEventHandler implements WebhookEventHandler {
+
+  private ObjectMapper objectMapper;
+
+  @Autowired
+  public BitbucketWehbookEventHandler(ObjectMapper objectMapper) {
+    this.objectMapper = objectMapper;
+  }
+
+  public boolean handles(String source) {
+    return source.equals("bitbucket");
+  }
+
+  public boolean shouldSendEvent(Event event) {
+    if(event.rawContent.isEmpty()) {
+      return false;
+    }
+
+    if (event.content.containsKey("hash") &&
+       event.content.get("hash").toString().startsWith("000000000")) {
+      return false;
+    }
+
+    return true;
+  }
+
+  public void handle(Event event, Map postedEvent) {
+    if (event.rawContent.isEmpty() || !event.content.containsKey("event_type")) {
+      return;
+    }
+
+    if (looksLikeBitbucketCloud(event)) {
+      handleBitbucketCloudEvent(event, postedEvent);
+    } else if (lookLikeBitbucketServer(event)) {
+      handleBitbucketServerEvent(event, postedEvent);
+    }
+  }
+
+  private boolean looksLikeBitbucketCloud(Event event) {
+    if (!event.content.containsKey("event_type")) {
+      return false;
+    }
+
+    String eventType = event.content.get("event_type").toString();
+    return (eventType == "repo:push" || eventType == "pullrequest:fulfilled");
+  }
+
+  private boolean lookLikeBitbucketServer(Event event) {
+    if (!event.content.containsKey("event_type")) {
+      return false;
+    }
+
+    String eventType = event.content.get("event_type").toString();
+    return (eventType == "repo:refs_changed" || eventType == "pr:merged");
+  }
+
+
+
+  private void handleBitbucketCloudEvent(Event event, Map postedEvent) {
+    String repoProject = "";
+    String slug = "";
+    String hash = "";
+    String branch = "";
+
+    BitbucketCloudEvent bitbucketCloudEvent = objectMapper.convertValue(postedEvent, BitbucketCloudEvent.class);
```

Comment:
Is it me or does this seem more complicated than the straight forward code before!?

Reply:
You're probably right. By moving from Groovy to Java we trade convenience for safety so thing are more verbose. I'm open to feedback if you have any suggestions for how to make this less mind-bending, ha!

Reply:
No, please no. :|

Reply:
> No, please no. :|

Exactly my reaction to Spinnaker being written in Java 😱